spring-security-modules/spring-security-web-boot-4/src/main/java/com/baeldung/apikeysecretauth/security/ApiKeySecretAuthenticationProvider.java

```diff
@@ -0,0 +1,51 @@
+package com.baeldung.apikeysecretauth.security;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import com.baeldung.apikeysecretauth.model.User;
+import com.baeldung.apikeysecretauth.repository.UserKeysRepository;
+import com.baeldung.apikeysecretauth.repository.model.UserKeysData;
+
+public class ApiKeySecretAuthenticationProvider implements AuthenticationProvider {
+    private final PasswordEncoder passwordEncoder;
+    private final UserKeysRepository userKeysRepository;
+
+    public ApiKeySecretAuthenticationProvider(PasswordEncoder passwordEncoder, UserKeysRepository userKeysRepository) {
+        this.passwordEncoder = passwordEncoder;
+        this.userKeysRepository = userKeysRepository;
+    }
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        final ApiKeySecretAuthenticationToken token = (ApiKeySecretAuthenticationToken) authentication;
+        final UserKeysData userKey = userKeysRepository.findById(token.getApiKey())
+            .orElseThrow(() -> new BadCredentialsException("Invalid API KEY"));
+
+        if (!passwordEncoder.matches(token.getApiSecret(), userKey.getApiSecret())) {
+            throw new BadCredentialsException("Invalid API SECRET");
+        }
+
+        final User user = new User(userKey.getAppUser());
+        return new ApiKeySecretAuthenticationToken(user, extractAuthorities(userKey));
+    }
+
+    @Override
+    public boolean supports(Class<?> authentication) {
+        return authentication.equals(ApiKeySecretAuthenticationToken.class);
+    }
+
+    private static Set<GrantedAuthority> extractAuthorities(UserKeysData userKey) {
+        return userKey.getPermissions()
+            .stream()
+            .map(p -> (GrantedAuthority) () -> p)
```

Comment:
This cast is a little ugly - see https://stackoverflow.com/a/22511918/1355930 for a way that looks neater with a function reference. `GrantedAuthority.class::cast`

Reply:
fixed